app/src/main/cpp/native-lib.cpp

```diff
@@ -0,0 +1,67 @@
+#include <jni.h>
+#include <android/log.h>
+#include <wallet.h>
+#include <string>
+#include <math.h>
+
+#include "../../../../jniLibs/wallet.h"
+#include <android/log.h>
+
+#define LOG_TAG "Tari Wallet"
+
+/**
+ * Log functions. Log example:
+ *
+ * int count = 5;
+ * LOGE("Count is %d", count);
+ * char[] name = "asd";
+ * LOGI("Name is %s", name);
+ */
+#define LOGE(...) __android_log_print(ANDROID_LOG_ERROR,    LOG_TAG, __VA_ARGS__)
+#define LOGW(...) __android_log_print(ANDROID_LOG_WARN,     LOG_TAG, __VA_ARGS__)
+#define LOGI(...) __android_log_print(ANDROID_LOG_INFO,     LOG_TAG, __VA_ARGS__)
+#define LOGD(...) __android_log_print(ANDROID_LOG_DEBUG,    LOG_TAG, __VA_ARGS__)
+
+extern "C"
+JNIEXPORT jstring JNICALL
+Java_com_tari_android_wallet_ui_activity_MainActivity_stringFromJNI(
+        JNIEnv *env,
+        jobject /* this */) {
+    std::string hello = "Message from JNI!";
+    return env->NewStringUTF(hello.c_str());
+}
+
+extern "C"
+JNIEXPORT void JNICALL
+Java_com_tari_android_wallet_ui_activity_MainActivity_changeCountJNI(
+        JNIEnv *pEnv,
+        jobject pThis,
+        jobject pObject) {
+    jclass objClass = pEnv->GetObjectClass(pObject);
+    jfieldID myFieldID = pEnv->GetFieldID(objClass, "count", "I");
+    // jint fieldVal = pEnv->GetIntField(pObject, myFieldID);
+    const int newCount = 8;
+    pEnv->SetIntField(pObject, myFieldID, newCount);
+
+    //std::string hello = "Hello from C++";
+    //return pEnv->NewStringUTF(hello.c_str());
+}
+
+extern "C"
+JNIEXPORT jstring JNICALL
+Java_com_tari_android_wallet_ui_activity_MainActivity_privateKeyStringJNI(
+        JNIEnv *pEnv,
+        jobject pThis) {
+    LOGI("This log is from JNI.");
+
+    TariPrivateKey *pk = private_key_generate();
+    ByteVector *bytes = private_key_get_bytes(pk);
+    unsigned int bytes_length = byte_vector_get_length(bytes);
+    std::string s = std::string("");
+    for (int i = 0; i <= bytes_length; i++)
+    {
+        unsigned int byte = byte_vector_get_at(bytes,i);
+        s += static_cast<char>(trunc(byte/2)); //should be converted to hex, div by 2 to ensure utf8 valid
+    }
+    return pEnv->NewStringUTF(s.c_str());
```

Comment:
This was part of sample test code provided.

Note that *pk and *bytes go out of scope after this return however the memory they are pointing to is not freed leading to a leak.

Reply:
@StriderDM I guess I'll just add before return:
```cpp
byte_vector_destroy(bytes); 
private_key_destroy(pk);
```

Reply:
This will work